dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/CommittingPartitionTracker.java

```diff
@@ -0,0 +1,304 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.kafka;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.KafkaAdapterFactory;
+import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.OffsetAndMetadata;
+import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.TopicPartition;
+
+/**
+ * Tracks the assignment of a partition to this consumer and commits message offsets to that partition
+ * <p>
+ * In addition to the function of {@link PartitionTracker}, this class additionally commits message offsets back to the kafka broker in response to a call to
+ * {@link #recordDone(long, Optional)}.
+ * <p>
+ * For performance, each completed record isn't committed immediately. Instead, this class tries to batch up completed records and commit them together. There are several
+ * parameters that control how often the message offset is committed:
+ * <ul>
+ * <li>{@code maxCommitBatchSize}: sets the maximum number of records to wait for before committing the offset</li>
+ * <li>{@code maxCommitBatchInterval}: sets the maximum time to wait after a {@link #recordDone(long, Optional)} is called before committing the offset</li>
+ * </ul>
+ */
+public class CommittingPartitionTracker extends PartitionTracker {
+
+    private static final TraceComponent tc = Tr.register(CommittingPartitionTracker.class);
+
+    private final KafkaAdapterFactory factory;
+    private final ScheduledExecutorService executor;
+    private final KafkaInput<?, ?> kafkaInput;
+    private final int maxCommitBatchSize;
+    private final Duration maxCommitBatchInterval;
+
+    /**
+     * Set of CompletedWork which has either not been committed or has been committed but the commit has not completed yet.
+     * <p>
+     * Any access to this must be synchronized
```

Comment:
synchronized on what? assume it is always `completedWork` for all of these variables?